### PR TITLE
NEXUS-8588: Format specific write policy

### PR DESCRIPTION
**Pull request by cstamas**
Some formats require the capability to override configured write policy. In case of Maven, IF configured write policy is ALLOW_ONCE, the effective write policy for repository metadata must be ALLOW, as every deploy (that only creates but does not updates artifacts) will attempt to update the metadata.

Issue
https://issues.sonatype.org/browse/NEXUS-8588

CI
http://bamboo.s/browse/NX3-OSSF286

### CODE REVIEW

**Comment by cstamas** [comment]
http://bamboo.s/browse/NX3-OSSF286-1 :+1:

**Comment by mosabua** [comment]
Nice ;-)

**Comment by adreghiciu** [comment]
+1

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacet.java [hunk 1]
```diff
@@ -53,6 +53,12 @@
   static String P_VERSION = "version";
 
   /**
+   * Registers format specific selector for {@link WritePolicy}. If not set, the {@link
+   * WritePolicySelector#DEFAULT} is used (returns the configured write policy).
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacet.java]: s/is used (returns the configured write policy)/is used which returns the configured write policy/

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java [hunk 2]
```diff
@@ -451,9 +456,9 @@ public Blob requireBlob(final BlobRef blobRef) {
     return blob;
   }
 
-  private void deleteBlob(final BlobRef blobRef, boolean checkWritePolicy) {
+  private void deleteBlob(final Asset asset, final BlobRef blobRef, boolean checkWritePolicy) {
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java]: Could just pass in the effective write policy here which avoids the need to pass in the asset and select the policy again. ie. `deleteBlob(final BlobRef blobRef, WritePolicy writePolicy) {` - clients would then select the policy or pass in `null` if the policy shouldn't be checked.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/WritePolicySelector.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.storage;
+
+/**
+ * Write policy selector.
+ *
+ * @since 3.0
+ */
+public abstract class WritePolicySelector
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/WritePolicySelector.java]: Any reason this isn't an interface? Interfaces can still contain static members which are concrete implementations of the interface

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/WritePolicySelector.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+{
+  /**
+   * Default instance of {@link WritePolicy} selector, that returns the configured {@link WritePolicy} unchanged.
+   */
+  public static final WritePolicySelector DEFAULT = new DefaultWritePolicySelector();
+
+  private static class DefaultWritePolicySelector
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/WritePolicySelector.java]: could make this an anonymous class, ie:
```
WritePolicySelector DEFAULT = new WritePolicySelector() { ...
```

**Comment by cstamas** [comment]
http://bamboo.s/browse/NX3-OSSF286-2 :+1:

**Comment by cstamas** [comment]
Merged as dd407fbf31097451412994190328224604884675

**Pull request closed without merging**